Add static ifconfig from NVRAM, enable VME, add NFSv4

hjunkes opened the pull request:
Uses rtems6 with libbsd. Extensively tested with MVME6100 and VMEbus.
Settings from NVRAM (gev) are evaluated according to Application Developers Guide. Alternatively dhcp can still be used. NFSv4 is usable and seems to work so far.
There are still conflicts with rtems/6/make/main.cfg (o-optimize, distclean-generic, clean-generic)
```
make[4]: Entering directory '/home/rtems/EPICS_TST/RTEMS_PR/epics-base/modules/normativeTypes/test/O.RTEMS-beatnik'
/home/rtems/MVME6100_WORK/rtems/6/make/main.cfg:44: warning: overriding recipe for target 'o-optimize'
/home/rtems/MVME6100_WORK/rtems/6/make/main.cfg:44: warning: ignoring old recipe for target 'o-optimize'
/home/rtems/MVME6100_WORK/rtems/6/make/main.cfg:72: warning: overriding recipe for target 'distclean-generic'
/home/rtems/MVME6100_WORK/rtems/6/make/main.cfg:72: warning: ignoring old recipe for target 'distclean-generic'
/home/rtems/MVME6100_WORK/rtems/6/make/main.cfg:76: warning: overriding recipe for target 'clean-generic'
/home/rtems/MVME6100_WORK/rtems/6/make/main.cfg:76: warning: ignoring old recipe for target 'clean-generic'
```

hjunkes committed:
Add static ifconfig from NVRAM, enable VME, add NFSv4

hjunkes committed:
set defines RTEMS_HAS_NETWORKING and RTEMS_HAS_POSIX hard coded for v6

hjunkes committed:
Add -DHAS_RTC flag (uC5282 has not for example)

hjunkes committed:
Add check for 6-freebsd-12 due to NFSv4

hjunkes committed:
Add check for 6-freebsd-12 due to rtems_telnetd_initialize

mdavidsaver commented on modules/libcom/RTEMS/posix/rtems_init.c:
```diff
@@ -1018,12 +1061,12 @@ POSIX_Init ( void *argument __attribute__((unused)))
-    printf("\n***** Initializing network (libbsd, dhcpcd) *****\n");
-    rtems_bsd_setlogpriority("debug");
+    printf("\n***** Initializing network (libbsd) *****\n");
+    if(0) rtems_bsd_setlogpriority("debug");
```
Is this PR intended for immediate consideration, or should it be marked as a draft?

hjunkes replied:
I would love it if that was seen as a "serious" next step and could be incorporated into the next release.

I am now using rtems6 with 6-freebsd-12. This gives support for NFSv4 which had been requested by some.
I am also now using a nexus ethernet driver with the MVME6100.

Beside the dhcpcd support the CPU can now also get its network configuration in the "old" way from the NVRAM. That was also requested by a few places. 

Also the VMEbus support is now back with rtems6 (I had this still excluded since still tests were missing from my side).

mdavidsaver replied:
I'm referring specifically to `if(0)`, which imo. is unsightly clutter.

Making the log level/priority configurable would be an improvement, with NVRAM storage being an obvious place to store such an option.  Though, I think that "debug" should remain the default so that new users have more indication of what is going on.

hjunkes committed:
Debug as default

mdavidsaver commented:
> Settings from NVRAM (gev) ...

Referencing myself, since I can't find an authoritative document at the moment.  The format of the VPD EEPROM used by Motload.  In two parts.  First the Vital Product Data at offset 0.  RTEMS BSPs read this early during boot. 

 https://github.com/mdavidsaver/qemu/blob/04ae8ce0a3da9467880064114fa370744a7811cf/hw/ppc/mvme3100.c#L138-L165  

Then, in a different format, from offset 0x1078 (4344) are user editable Global Environment Variables.  This is what `setBootConfigFromNVRAM()` is reading when `defined(HAVE_MOTLOAD)`.

https://github.com/mdavidsaver/qemu/blob/04ae8ce0a3da9467880064114fa370744a7811cf/hw/ppc/mvme3100.c#L277-L281

For an mvme3100, the VPD EEPROM is documented as 8 KB in total.

anjohnson commented:
@mdavidsaver Some corrections to your VPD format:

The "eyecatcher" that identifies the start of the VPD data is actually an 8-byte string "MOTOROLA" with no trailing nil (you have "MOTLOAD"). This applies to the MVME2100, 3100, 5100, 5500, and 6100. For the MVME2500 it is "EMERSON ", again with no nil.

For MAC addresses (packet type 08) if present the 7th byte of the packet data is actually the instance number of the interface, so this dump of a VPD that VxWorks copied into RAM shows 3 MAC addresses with instances 0, 1 and 2, starting at 0x00301bb2:
```
iocanj> d sysVpdPkts, 0x180,1
0x00301b60:        0f 04 00 03 00 00 06 05 03 f9 40 aa 01 01  *   .........@...*
0x00301b70:  0d 4d 56 4d 45 33 31 30 30 2d 31 31 35 32 02 0c  *.MVME3100-1152..*
0x00301b80:  30 31 2d 57 33 38 39 33 46 30 32 44 03 07 37 36  *01-W3893F02D..76*
0x00301b90:  35 32 34 36 30 0e 0f ff ff ff ff ff ff ff ff ff  *52460...........*
0x00301ba0:  01 ff 01 ff ff ff 09 04 38 35 34 30 0a 04 d1 e0  *........8540....*
0x00301bb0:  35 75 08 07 00 01 af 19 54 32 00 08 07 00 01 af  *5u......T2......*
0x00301bc0:  19 54 33 01 08 07 00 01 af 19 54 34 02 0b 0c 00  *.T3.......T4....*
0x00301bd0:  01 7e 22 10 02 02 10 10 00 64 08 ff ff ff ff ff  *.~"......d......*
```
You can tell whether an instance number is included or not based on the packet size, which apparently may be either 6 or 7, although it looks like they are always 7 on our boards if used (the MVME2100 and 2700 get their MAC address from the network interface).

VPD packet type 04 is "Product Configuration Options" and is basically a bit-mask; 07 is "Reference Clock Speed (in HZ)"

mdavidsaver commented:
> is actually an 8-byte string "MOTOROLA"

Ah, my comment is wrong.  The code is below writes the correct string.

https://github.com/mdavidsaver/qemu/blob/04ae8ce0a3da9467880064114fa370744a7811cf/hw/ppc/mvme3100.c#L249

I have also now noticed a couple of samples tucked away in the rtems repository.

https://github.com/RTEMS/rtems/blob/a316a9ddaeaa8f6316b2a2d29ca82b3ad40d2d22/bsps/powerpc/shared/start/vpd.c#L234-L322

anjohnson commented:
If you want another one, here's an MVME2500:
```
iocanj1> d &sysVpd, 0x80,1
0x0033c990:  45 4d 45 52 53 4f 4e 20 00 80 03 07 44 30 30 37  *EMERSON ....D007*
0x0033c9a0:  41 43 41 0a 04 6d 4f 47 4c 0f 04 05 04 21 ff 01  *ACA..mOGL....!..*
0x0033c9b0:  14 4d 56 4d 45 32 35 30 30 2d 30 31 30 38 30 31  *.MVME2500-010801*
0x0033c9c0:  30 31 53 20 20 02 0d 30 31 30 36 38 35 34 4c 30  *01S  ..0106854L0*
0x0033c9d0:  35 57 20 20 09 05 50 32 30 31 30 05 04 2f af 08  *5W  ..P2010../..*
0x0033c9e0:  00 06 04 05 f5 e1 00 08 07 ec 9e cd 23 20 7f 00  *............# ..*
0x0033c9f0:  08 07 ec 9e cd 23 20 80 01 08 07 ec 9e cd 23 20  *.....# .......# *
0x0033ca00:  81 02 04 02 ff b3 ff ff ff ff ff ff ff ff ff ff  *................*
value = 0 = 0x0
iocanj1> vpdShow          
 Type = 0x03 - Serial Number
 Size = 0x07
Value = D007ACA

 Type = 0x0a - EEPROM CRC
 Size = 0x04
Value = 6D4F474C

 Type = 0x0f - VPD Revision
 Size = 0x04
Value = 05 04 21 FF 

 Type = 0x01 - Product Identifier
 Size = 0x14
Value = MVME2500-01080101S  

 Type = 0x02 - Factory Assembly Number
 Size = 0x0d
Value = 0106854L05W  

 Type = 0x09 - Microprocessor Type
 Size = 0x05
Value = P2010

 Type = 0x05 - Internal Clock Speed (Hz)
 Size = 0x04
Value = 800000000

 Type = 0x06 - External Clock Speed (Hz)
 Size = 0x04
Value = 100000000

 Type = 0x08 - Ethernet Address
 Size = 0x07
Value = EC:9E:CD:23:20:7F  # 0

 Type = 0x08 - Ethernet Address
 Size = 0x07
Value = EC:9E:CD:23:20:80  # 1

 Type = 0x08 - Ethernet Address
 Size = 0x07
Value = EC:9E:CD:23:20:81  # 2

 Type = 0x04 - Product Configuration Options
 Size = 0x02
Value = FF B3 

 Type = 0xff - Termination Marker

value = 0 = 0x0
```

hjunkes committed:
Re-enable the forgotten NTPSupport (osdTimeRegister()) for the LEGACY_STACK.

hjunkes commented:
Chris had pointed out to me that the NTP support in EPICS-7 was gone. That was my mistake, I had simply "forgotten" that. Sorry for that. I have it now again for the LEGACY_STACK activated. Tested with RTEMS 5.1 on PPC.
I'm still trying to activate the service for the new libbsd stack as well. Could be a topic for our meeting at the end of October.

anjohnson commented:
Hello Heinz, one improvement I added to the VxWorks osdTime.c in the latest Base release was to not start the EPICS NTP sync task if the OS-provided sync task is running and the OS clock seems to be set (i.e. the current clock time is after the build time of the code), so two sync tasks won't try to fight if they happen to be synchronized to a different NTP source. The user can override this by setting the `EPICS_TS_FORCE_NTPTIME` environment variable, and in that case the `timeRegister()` routine will attempt to shut down the OS sync task if it's running, so EPICS should then take over the job of keeping the OS time in sync.

goetzpf committed:
Fix generation of mvme2100 bootable files.

The recipe how to create a bootable file was taken from here:

  https://docs.rtems.org/branches/master/user/bsps/bsps-powerpc.html#boot-image-generation

This is at least needed for the mvme2100 CPU on order to boot via PPCBUG.

goetzpf committed:
Fix mvme2100 memory problem.

This is a problem due to the fact that the mvme2100 CPU has only 32MB of
memory. The problem and the solution is discussed here:

https://www.mail-archive.com/users@rtems.org/msg03396.html

goetzpf committed:
Allow tftp and nfs for fetching the startup script.

Note: In order to specify the startup script to be fetched via tftp from the
BOOTP_HOST (the host where the kernel was loaded from) specify this as path
("Argument File Name" in PPCBUG "niot" command):

  /TFTP/BOOTP_HOST:<path to st.cmd>

If the startup script should load files also via tftp, do this in the script:

  cd("/TFTP/BOOTP_HOST:<path>/")

Ensure that this path ends with a slash "/".

hjunkes committed:
Introduce Definition HAS_RTC
to control CONFIGURE_APPLICATION_NEEDS_RTC_DRIVER config file

hjunkes committed:
Merge branch 'epics-base:7.0' into rtems6

hjunkes committed:
rtems_telnetd_initialize only for PPC

hjunkes committed:
Merge branch 'rtems6' of https://github.com/hjunkes/epics-base into rtems6

hjunkes committed:
Add missing string.h

hjunkes committed:
Disable bootp for pc686 on rtems

hjunkes committed:
Do not start NTP-sync, NFS-mount, etc. if only loopback-if is up

hjunkes committed:
epicsThreadSetPriority set prio even if non epics thread

mdavidsaver committed:
ringPointerTest in thread

hjunkes committed:
Add sleeps to very short test

hjunkes committed:
removed sleep in test to see if its fails

hjunkes committed:
Add delay in perl test area to exit

hjunkes committed:
Merge branch 'epics-base:7.0' into rtems6

hjunkes committed:
Merge branch 'RTEMS-mvme2100-fixes-7.0' into rtems6

hjunkes committed:
Add / modify / test Rtems mvme2100 fixes 7.0 #299

hjunkes committed:
Merge branch 'ringPointerTest-prio' into rtems6

anjohnson commented on modules/libcom/src/osi/os/RTEMS/osdTime.cpp:
```diff
@@ -156,17 +174,25 @@ double rtemsTicksPerSecond_double, rtemsTicksPerTwoSeconds_double;
+/* this does not work if we use DHCP. TBD! */
+/* We should check that the interface is up ... */
+#if __RTEMS_MAJOR__ < 5
+    if (rtems_clock_get_ticks_per_second != 0)
```
TBD, needed by qemu-pc686

anjohnson commented on modules/libcom/src/osi/os/RTEMS/osdTime.cpp:
```diff
@@ -21,28 +23,37 @@
     NTPTime_Init(100);
     ClockTime_Init(CLOCKTIME_SYNC);
-
+#else
+printf(" OsdTimeRegister Not implemented yet for new libbsd ...\n");
+#endif
```
Not implemented yet.

hjunkes replied:
void osdTimeRegister(void)
{
    /* Init NTP first so it can be used to sync ClockTime */
#ifdef RTEMS_LEGACY_STACK
    NTPTime_Init(100);
    ClockTime_Init(CLOCKTIME_SYNC);
#else
printf(" osdTimeRegister: not implemented for non RTEMS_LEGACY_STACK \n");
#endif
    osdMonotonicInit();
}

anjohnson commented on modules/libcom/src/osi/os/RTEMS/osdTime.cpp:
```diff
@@ -83,7 +94,12 @@ void osdNTPInit(void)
 {
     struct sockaddr_in myAddr;
 
+#if __RTEMS_MAJOR__ > 4
+    ntpSocket = socket( AF_INET, SOCK_DGRAM, IPPROTO_UDP );
+#else
     ntpSocket = socket (AF_INET, SOCK_DGRAM, 0);
+#endif
```
Check whether this `#if` is needed, `IPPROTO_UDP ` should work on all RTEMS versions, explain why it's needed in a comment if it doesn't.

hjunkes replied:
I have removed it, it is probably not necessary? In other implementations e.g. vxWorks the socket is not created at all, and also in  rtems_bsdnet_get_ntp() a socket is created if it does not exists already. Probably it is used here to be able to delete old answers of the NTP servers before the request.

I also took the opportunity to add a little information about the socketNr and the NTP server in osdNTPReport:

DYCRVME> NTPTime_Report
NTP driver is synchronized with server
Syncronization interval = 60.0 seconds
Last synchronized at 2023-01-05 13:38:40.114914
Current OS tick rate = 100 Hz
Measured tick rate = 0.000 Hz
osdNTPReport: ntpSocket used : 28
osdNTPReport: NTP Server = 10.0.0.1

anjohnson commented on modules/libcom/RTEMS/setBootConfigFromNVRAM.c:
```diff
@@ -23,12 +21,45 @@
 char *env_nfsMountPoint;
 
-extern char* rtems_bsdnet_bootp_cmdline;
+#ifndef RTEMS_LEGACY_STACK
```
Create macros here and in a `#else` case to abstract the accesses to these members below, reducing the number of `#if` conditionals further down.

anjohnson reviewed (commented):
Core Group, 2023-01-04: A few comments inline.

hjunkes committed:
osdNTPReport info added, rtems version checks on socket create removed

hjunkes committed:
Test if it now compiles with rtems 4.

hjunkes committed:
Another Test for pc686 libbsd (not LEGACY_STACK)

hjunkes committed:
respect -DRTEMS_LEGACY_STACK also for rtems 4.x

hjunkes committed:
removed include of rtems-internal-header
rtems_clock_get_ticks_per_second() does never return 0

hjunkes committed:
Merge branch 'epics-base:7.0' into rtems6

hjunkes committed:
Merge branch '7.0' into rtems6

hjunkes committed:
Add -isystems to CONFIG due to conflict with rtems dblink.h

hjunkes committed:
Add nfsv4 definition to xilinx_zynq_a9_qemu config file, adapt LDFLAGS

hjunkes committed:
Added fixed address settings for qemu in setBootConfigFromNVRAM.c

hjunkes committed:
No NFS mount attempt if localFilesystem exists

hjunkes committed:
Use new linker symbols

hjunkes committed:
Added 1ms-per-tick configuration
The analysis behind this change is:
https://gitlab.com/nsf-noirlab/gemini/rtsw/epics-base/epics-base/-/issues/30

hjunkes committed:
Configuration definition EPICS_IS_QEMU should not be public

hjunkes committed:
Cleanup

hjunkes committed:
Add support for mvme5500

hjunkes committed:
Merge branch 'epics-base:7.0' into rtems6

hjunkes committed:
Add Goetz Pfeiffers patch for unknown reference bsp_section_data_end
on other boards than MVME2500 (qoriq)

hjunkes committed:
Merge branch 'epics-base:7.0' into rtems6

hjunkes committed:
Add HAS_EDATA Flag for local_filesystem

hjunkes committed:
Merge branch 'rtems6' of https://github.com/hjunkes/epics-base into rtems6

hjunkes committed:
Another CPU with _edata?

hjunkes committed:
I used the wrong identifier for the BSP.

hjunkes committed:
Also uC5282 has _edata

hjunkes commented:
Considering Chris's PR (https://github.com/epics-base/epics-base/pull/375), I am withdrawing this PR to make it a little easier to integrate everything.

hjunkes commented:
Considering Chris's PR (https://github.com/epics-base/epics-base/pull/375), I am withdrawing this PR to make it a little easier to integrate everything.
